preprocessing.py

```diff
@@ -6,33 +6,36 @@
 import os
 
 
-def one_doc_preprocess(ar): #This is dataset-specific code. I think that you should name this method as "cacm_doc_preprocess"
+def one_doc_preprocess(ar, list_of_arguments): 
     tree = ET.parse(ar)
     root = tree.getroot()
     stop_words = set(stopwords.words('english'))
     translator = str.maketrans('', '', string.punctuation)
     ps = PorterStemmer()
-    a = [elem.tag for elem in root.iter()] #for what reasons you created a-variable and didn't use it?
-    for docid in root.iter("DOCID"): #Do I understand correctly that one XML is containing only one document (one DOCID, title, body) ?
-        name = docid.text.replace(" ", "") #what if document doesn't contain name, title or other fields? I think that you should prevent strange errors with format checking 
-                                           # like if name == None then exit with message "document has invalid format"
+    for docid in root.iter("DOCID"): #Do I understand correctly that one XML is containing only one document (one DOCID, title, body) ? Yes
+        name = docid.text.replace(" ", "")
+        if docid == None:
+            raise ValueError("document has invalid format")
     dest = "preprocessed_files"
     if not os.path.exists(dest):
         os.mkdir(dest)
-    for title in root.iter("TITLE"):
-        s = title.text.lower().replace("-", ' ').translate(translator).split() #see comment above
-    s = [w for w in s if w not in stop_words] #using of list generator is not a good idea, if document will have big size then this code may fail with out of memory
-                                              # you can rewrite it to for-loop with if statement in body like:
-                                              # for word in s:
-                                              #     if word not in stop_words:
-                                              #         ...
-    for w in s:
-        print(ps.stem(w), file=open(dest + '/' + name, "a")) #unfortunately I haven't cacm-dataset in such format and I cannot see output of script
-                                                             #but what about spaces between words? ps.stem(w) serialization gives word with spaces?
+    for argument in list_of_arguments:
+        s = []
+        for current_field in root.iter(argument):
+            s = current_field.text.lower().replace("-", ' ').translate(translator).split() #see comment above
+            if current_field == None:
+                raise ValueError("document has invalid format")
+        s1 = []
+        for word in s:
```

Comment:
why you cannot write like this or similar?

with open(filename, 'w') as output_file:
-- for word in s:
---- if word not in stop_words:
------ print(ps.stem(w), output_file)

You should understand that "print(..., file=open(...))" opens file foreach word in list "s" and writes only one word. Moreover, if you use such construction you don't close file after printing. This is a bug because file descriptor will not be released.

Also why you use "s1"-list when you can filter words and print them directly in one code block? More lists more memory your application requires. If this list is not necessary for logic of code then it should be eliminated.